source/lib-intro.tex

```diff
@@ -2652,7 +2733,7 @@
 paragraph.
 \item
 if an incomplete type\iref{basic.types} is used as a template
-argument when instantiating a template component, unless specifically
+argument when instantiating a template component or evaluating a concept, unless specifically
```

Comment:
(Not editorial) I find it odd that an incomplete type would cause undefined behavior; it's a compile-time failure and should be ill-formed (possibly no diagnostic required).

Reply:
LWG has been addressing a similar problem with the type traits to require implementations to diagnose precondition violations. There's currently no technology to do so with concepts, but this at least leaves the avenue open for future changes.

Reply:
I'm complaining about the fact that this situation is in the "undefined behavior" category, not in a (separate) category "ill-formed, no diagnostic required". The latter would leave avenues open, too, but would also permit implementations to diagnose and refuse e.g. incomplete types today. And, maybe there is even a conflict here if some preconditions says "shall be a complete type" (i.e. compile-time diagnostic), yet this section says "undefined behavior" (i.e. only bad if you actually evaluate it, which seems a misguided idea for a compile-time violation).

Reply:
I agree. I've made a note to file an LWG issue to adjust this wording to IFNDR.

Reply:
Thank you.

Reply:
This is [LWG 3142](https://wg21.link/lwg3142).